modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryMemoryTracker.java

```diff
@@ -54,47 +66,63 @@
     QueryMemoryTracker(H2MemoryTracker parent, long maxMem, long blockSize) {
         assert maxMem > 0;
 
+        // TODO GG-18629 - get from configuration.
+        failOnMemLimitExceed = !Boolean.getBoolean(IgniteSystemProperties.IGNITE_SQL_USE_DISK_OFFLOAD);
         this.parent = parent;
         this.maxMem = maxMem;
         this.blockSize = blockSize;
     }
 
     /** {@inheritDoc} */
-    @Override public void reserve(long size) {
-        if (size == 0)
-            return;
-
-        assert size > 0;
+    @Override public boolean reserved(long size) {
+        assert size >= 0;
```

Comment:
Why is fast-path for (size==0) removed?
Seems, it can be very helpful for " reserved(0) " call in needSpillToDisk.

Reply:
We need to return boolean value if we have or don't have a free memory. Anyway we need to check the current reserved memory value.